### PR TITLE
fix issue with duplicate completion in MultipleConnectAsync

### LINKED ISSUE #43091: Test failure: System.Net.Sockets.Tests.ConnectEap.Connect_Dns_Success
`Connect_Dns_Success` socket test is failing for the EAP case with the error of the [event handler](https://github.com/dotnet/runtime/blob/f2d51b5e15808d2c358a59f24af4fad4a9953656/src/libraries/System.Net.Sockets/tests/FunctionalTests/SocketTestHelper.cs#L296) being called twice.

```
System.Net.Sockets.Tests.ConnectEap.Connect_Dns_Success(listenAt: 127.0.0.1) [FAIL]
      System.InvalidOperationException : An attempt was made to transition a task to a final state when it had already completed.
      Stack Trace:
        /_/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/TaskCompletionSource_T.cs(202,0): at System.Threading.Tasks.TaskCompletionSource`1[[System.Boolean, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].SetResult(Boolean result)
        /_/src/libraries/System.Net.Sockets/tests/FunctionalTests/SocketTestHelper.cs(296,0): at System.Net.Sockets.Tests.SocketHelperEap.<>c__DisplayClass15_0`1[[System.Boolean, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].<InvokeAsync>b__0(Object _, SocketAsyncEventArgs e)
        /_/src/libraries/System.Net.Sockets/tests/FunctionalTests/SocketTestHelper.cs(303,0): at System.Net.Sockets.Tests.SocketHelperEap.InvokeAsync[Boolean](Socket s, Func`2 getResult, Func`2 invoke)
        /_/src/libraries/System.Net.Sockets/tests/FunctionalTests/SocketTestHelper.cs(241,0): at System.Net.Sockets.Tests.SocketHelperEap.ConnectAsync(Socket s, EndPoint endPoint)
        /_/src/libraries/System.Net.Sockets/tests/FunctionalTests/SocketTestHelper.cs(329,0): at System.Net.Sockets.Tests.SocketTestHelperBase`1[[System.Net.Sockets.Tests.SocketHelperEap, System.Net.Sockets.Tests, Version=5.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51]].ConnectAsync(Socket s, EndPoint endPoint)
        /_/src/libraries/System.Net.Sockets/tests/FunctionalTests/Connect.cs(59,0): at System.Net.Sockets.Tests.Connect`1.<Connect_Dns_Success>d__3[[System.Net.Sockets.Tests.SocketHelperEap, System.Net.Sockets.Tests, Version=5.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51]].MoveNext()
```

The first occurrence of this failure was on CI for PR #42919 that worked around fixing `MultipleConnectAsync` locking logic.
Failed 18 times in innerloop in the last 10h after the PR was merged.

Kusto: [link](https://dataexplorer.azure.com/clusters/engsrvprod/databases/engineeringdata?query=H4sIAAAAAAAAA22QT2vCQBDF736KRy62EDV6ayEFWxEsVGgVepRpdmrWJLthZ4IIfvhu1PZUWNg/85t57+2WRT9YuloFgzMO3jpU1pncOscBnz5UK+VG4N3fZWX+Q1/914WK+8rE+rHkwFhaZ6Vkg6cchpTVNnw3y2bZKHsYTZE9Zv26H4CcwfbUMkQpqBytlkg2J4l64zXrOLkib6ylN8hzJC8+6ha6WzjZbbqiYJEbdA3UQ8Ml2Xo4wGRybRahPaPwTsk6QTJ3II0SreJIgoYMQz00kBOrNsYhKEnVPxK+raO6N6jcx3OwipIMqA5M5hTnNm3Nyia6PaMN/hD9YdMHYpNi0QXqZ6a3FCneO+54TQ1P019zac8X1TZQEc/zsO8adiqxTkVp3YVO8RwNFmV6+bEfECMibcMBAAA=)

### PR DESCRIPTION
**Pull request by geoffkizer**
Fixes #43091 

This was introduced in PR #42919, which solved part of the issue here but didn't thread the sync handling all the way down. This PR does that.

@MihaZupan can you take a look?

### CODE REVIEW

**Review by wfurt — APPROVED** [review]
LGTM

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs [hunk 1]
```diff
@@ -685,8 +685,15 @@ internal void FinishWrapperConnectSuccess(Socket? connectSocket, int bytesTransf
             if (SocketsTelemetry.Log.IsEnabled()) LogBytesTransferEvents(connectSocket?.SocketType, SocketAsyncOperation.Connect, bytesTransferred);
 
             // Complete the operation and raise the event.
-            ExecutionContext? context = _context; // store context before it's cleared as part of completing the operation
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs:688]: This was just dead code before?

Comment (stephentoub) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs:688]: Oh, nevermind, diff view confused me.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs [hunk 2]
```diff
@@ -685,8 +685,15 @@ internal void FinishWrapperConnectSuccess(Socket? connectSocket, int bytesTransf
...
             Complete();
+        }
+
+        internal void FinishWrapperConnectAsyncSuccess(Socket? connectSocket, int bytesTransferred, SocketFlags flags)
+        {
+            ExecutionContext? context = _context; // store context before it's cleared as part of completing the operation
+
+            FinishWrapperConnectSyncSuccess(connectSocket, bytesTransferred, flags);
+
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs:696]: Doesn't this change mean we're no longer calling OnCompletedInternal for sync completions?

Comment (geoffkizer) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs:696]: Yes, we're no longer calling OnCompletedInternal for sync completions. That's the source of the bug -- we were both returning pending=false and calling the OnCompleted callback. Now we just return pending=false.

OnCompletedInternal doesn't log for the multi connect case anyway, there's an explicit check for it. I believe how it works is that it logs per connection attempt.

@MihaZupan Does that all sound right?

Comment (MihaZupan) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs:696]: Yes, for the Socket's connect telemetry we're only logging connection attempts, not the parent MultiConnect operation as a whole

**Pull request merged**